sphinx_gallery/directives.py

```diff
@@ -144,21 +144,41 @@ def has_backrefs(obj):
                 gallery_dirs = config.sphinx_gallery_conf["gallery_dirs"]
                 if not isinstance(gallery_dirs, list):
                     gallery_dirs = [gallery_dirs]
-                dirs = [
-                    (e, g)
-                    for e, g in zip(examples_dirs, gallery_dirs)
-                    if (obj.find(e) != -1)
-                ]
-                if len(dirs) != 1:
+
+                gal_matches = []
+                for e, g in zip(examples_dirs, gallery_dirs):
+                    e = Path(src_dir, e).resolve(strict=True)
+                    g = Path(src_dir, g).resolve(strict=True)
+                    try:
+                        gal_matches.append((path.relative_to(e), g))
+                    except ValueError:
+                        continue
+                # `path` inside one `examples_dirs`
+                if (n_match := len(gal_matches)) == 1:
+                    ex_parents, target_dir = (
+                        gal_matches[0][0].parents,
+                        gal_matches[0][1],
+                    )
+                # `path` inside several `examples_dirs`, take the longest match
+                elif n_match > 1:
+                    ex_parents, target_dir = max(
+                        [(match[0].parents, match[1]) for match in gal_matches],
+                        key=lambda x: len(x[0]),
+                    )
```

Comment:
I think this may have been what the old `len(dirs)>1` was testing before. The only case where I think this could come about is e.g.,

file: `/a/b/c/d/file.py`
examples dirs (abs path):
* `/a/b/`
* `/a/b/c/d` (i.e. you have a gallery within a sub-sub-folder of another gallery - this *may* be done if someone wanted to use more nesting for their galleries)

In this case, I think it makes sense that we just use the longest matching example dir.

Reply:
Hmm I don't think that is the same. E.g., `examples_dirs` in mpl would be e.g., `../galleries/examples`, `../galleries/plot_types/`. These folders have sub-gallaries/sub-folders, but these subfolderes do not need to be added to `examples_dirs` as SG auto checks for 1 level of sub-folders. 

It's only in the case where you want 2 levels of sub-folders e.g., you also have `../galleries/examples/animation/moreanimation/` in your `examples_dirs`.

It would be unusual case. And I don't even know if I can be bothered adding a test for this... :thinking: